### PR TITLE
[Clang] Fix Handling of Init Capture with Parameter Packs in LambdaScopeForCallOperatorInstantiationRAII

### PR DESCRIPTION
**Pull request by LYP951018**
This PR addresses issues related to the handling of `init capture` with parameter packs in Clang's `LambdaScopeForCallOperatorInstantiationRAII`.

Previously, `addInstantiatedCapturesToScope` would add `init capture` containing packs to the scope using the type of the `init capture` to determine the expanded pack size. However, this approach resulted in a pack size of 0 because `getType()->containsUnexpandedParameterPack()` returns `false`. After extensive testing, it appears that the correct pack size can only be inferred from `getInit`.

But `getInit` may reference parameters and `init capture` from an outer lambda, as shown in the following example:

```cpp
auto L = [](auto... z) {
    return [... w = z](auto... y) {
        // ...
    };
};
```

To address this, `addInstantiatedCapturesToScope` in `LambdaScopeForCallOperatorInstantiationRAII` should be called last. Additionally, `addInstantiatedCapturesToScope` has been modified to only add `init capture` to the scope. The previous implementation incorrectly called `MakeInstantiatedLocalArgPack` for other non-init captures containing packs, resulting in a pack size of 0.

### Impact

This patch affects scenarios where `LambdaScopeForCallOperatorInstantiationRAII` is passed with `ShouldAddDeclsFromParentScope = false`, preventing the correct addition of the current lambda's `init capture` to the scope. There are two main scenarios for `ShouldAddDeclsFromParentScope = false`:

1. **Constraints**: Sometimes constraints are instantiated in place rather than delayed. In this case, `LambdaScopeForCallOperatorInstantiationRAII` does not need to add `init capture` to the scope.
2. **`noexcept` Expressions**: The expressions inside `noexcept` have already been transformed, and the packs referenced within have been expanded. Only `RebuildLambdaInfo` needs to add the expanded captures to the scope, without requiring `addInstantiatedCapturesToScope` from `LambdaScopeForCallOperatorInstantiationRAII`.

### Considerations

An alternative approach could involve adding a data structure within the lambda to record the expanded size of the `init capture` pack. However, this would increase the lambda's size and require extensive modifications.

This PR is a prerequisite for implmenting https://github.com/llvm/llvm-project/issues/61426

### CODE REVIEW

**Comment by LYP951018** [comment]
ping~

File: clang/lib/Sema/SemaConcept.cpp [hunk 1]
```diff
@@ -712,22 +712,30 @@ bool Sema::addInstantiatedCapturesToScope(
...
     if (!CapturePattern.capturesVariable()) {
       Instantiated++;
       continue;
     }
-    const ValueDecl *CapturedPattern = CapturePattern.getCapturedVar();
+    ValueDecl *CapturedPattern = CapturePattern.getCapturedVar();
+
+    if (!CapturedPattern->isInitCapture()) {
+      continue;
+    }
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaConcept.cpp:728]: When is `CapturedPattern` not an initCapture?

Comment (LYP951018) [reply clang/lib/Sema/SemaConcept.cpp:728]: captures like
```cpp
[a]() {}
```

Comment (cor3ntin) [reply clang/lib/Sema/SemaConcept.cpp:728]: Hum, nvm, I misread that code , I failed to realize the assert L715 was in a different scope, so I was confused.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by LYP951018** [comment]
Thanks for the review ;)

File: clang/lib/Sema/SemaConcept.cpp [hunk 2]
```diff
@@ -712,22 +712,30 @@ bool Sema::addInstantiatedCapturesToScope(
...
+
     if (!CapturedPattern->isParameterPack()) {
       AddSingleCapture(CapturedPattern, Instantiated++);
     } else {
       Scope.MakeInstantiatedLocalArgPack(CapturedPattern);
-      std::optional<unsigned> NumArgumentsInExpansion =
-          getNumArgumentsInExpansion(CapturedPattern->getType(), TemplateArgs);
+      SmallVector<UnexpandedParameterPack, 2> Unexpanded;
+      SemaRef.collectUnexpandedParameterPacks(
+          dyn_cast<VarDecl>(CapturedPattern)->getInit(), Unexpanded);
```
Comment (zwuis) [inline clang/lib/Sema/SemaConcept.cpp:736]: Why not use `cast`?

**Pull request merged**